triedb/pathdb/states.go

```diff
@@ -43,9 +49,528 @@ func (c *counter) report(count metrics.Meter, size metrics.Meter) {
+			delete(s.accountData, accountHash)
+		}
+		if _, ok := s.storageData[accountHash]; ok {
+			// Looping through the nested map may cause slight performance degradation.
+			// However, since account destruction is no longer possible after the cancun
+			// fork, this overhead is considered acceptable.
+			for _, val := range s.storageData[accountHash] {
+				delta -= 2*common.HashLength + len(val)
+				storageOverwrites.add(2*common.HashLength + len(val))
+			}
+			delete(s.storageData, accountHash)
+		}
+		// Keep track of whether the account has already been marked as destructed.
+		// This additional marker is useful for undoing the merge operation.
+		_, exist := s.destructSet[accountHash]
+		destructs = append(destructs, destruct{
+			Hash:  accountHash,
+			Exist: exist,
+		})
+		if exist {
+			continue
+		}
+		delta += common.HashLength
+		s.destructSet[accountHash] = struct{}{}
+	}
+	s.journal.add(destructs)
+
+	// Apply the updated account data
+	for accountHash, data := range other.accountData {
+		if origin, ok := s.accountData[accountHash]; ok {
+			delta += len(data) - len(origin)
+			accountOverwrites.add(common.HashLength + len(origin))
+		} else {
+			delta += common.HashLength + len(data)
+		}
+		s.accountData[accountHash] = data
+	}
+	// Apply all the updated storage slots (individually)
+	for accountHash, storage := range other.storageData {
+		// If storage didn't exist (or was deleted) in the set, overwrite blindly
+		if _, ok := s.storageData[accountHash]; !ok {
+			// To prevent potential concurrent map read/write issues, allocate a
+			// new map for the storage instead of claiming it directly from the
+			// passed external set. Even after merging, the slots belonging to the
+			// external state set remain accessible, so ownership of the map should
+			// not be taken, and any mutation on it should be avoided.
+			slots := make(map[common.Hash][]byte, len(storage))
+			for storageHash, data := range storage {
+				slots[storageHash] = data
+				delta += 2*common.HashLength + len(data)
+			}
+			s.storageData[accountHash] = slots
+			continue
+		}
+		// Storage exists in both local and external set, merge the slots
+		slots := s.storageData[accountHash]
+		for storageHash, data := range storage {
+			if origin, ok := slots[storageHash]; ok {
+				delta += len(data) - len(origin)
+				storageOverwrites.add(2*common.HashLength + len(origin))
+			} else {
+				delta += 2*common.HashLength + len(data)
+			}
+			slots[storageHash] = data
+		}
+	}
+	accountOverwrites.report(gcAccountMeter, gcAccountBytesMeter)
+	storageOverwrites.report(gcStorageMeter, gcStorageBytesMeter)
+	s.clearCache()
+	s.updateSize(delta)
+}
+
+// revert takes the original value of accounts and storages as input and reverts
+// the latest state transition applied on the state set.
+func (s *stateSet) revert(accountOrigin map[common.Hash][]byte, storageOrigin map[common.Hash]map[common.Hash][]byte) {
```

Comment:
Btw, it's important to mention that: the `storageOrigin` contains all the mutated storage slots along with their original value;

If the account was destructed in the earlier state transition, all the storage slots belonging to this account were deleted as well; here, in the `storageOrigin`, all these storage slots will be present along with their original value;


It's quite different with `state.merge`, in which `storageData` ONLY represents the mutated slots of this account; If the account was destructed, it will be in the `destructSet`, and the corresponding storage slots (deleted) won't be present in `storageData`!

Reply:
it's no longer relevant!